### PR TITLE
[0.1] Delete machines

### LINKED ISSUE #985: [0.1] Deleting Cluster does not delete
/kind bug

**What steps did you take and what happened:**

1. Create a cluster in the cluster API
2. Add a Machine or MachineDeployment
3. Wait for all objects to become available
4. `kubectl delete` the cluster
5. The cluster gets a deleted-at, but nothing else gets deleted

**What did you expect to happen:**

The Machines, then MachineDeployments should be marked to be deleted, then the cluster deleted once their resources have been completely removed

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Environment:**

- Cluster-api version: [master](https://github.com/liztio/cluster-api/commit/2ec456177c0e8f0a903f4e746d44baaae54cc591)
- Minikube/KIND version: 0.1.0-alpha
- Kubernetes version: (use `kubectl version`): 1.13
- OS (e.g. from `/etc/os-release`): Ubuntu Bionic

### LINKED ISSUE #1190: [0.1] MachineDeployments, MachineSets, and Machines should always have a Cluster owner reference if a Cluster exists
/kind bug

**What steps did you take and what happened:**
- Create a Cluster
- Create a MachineDeployment, MachineSet, or Machine and manually specify an ownerReference to some arbitrary resource that is not a Cluster
- The created resource does not end up with an ownerReference pointing at the Cluster

**What did you expect to happen:**
- There should be an ownerReference pointing at the Cluster

**Anything else you would like to add:**
The MachineDeployment, MachineSet, and Machine controllers only add an ownerReference to the Cluster if there are 0 ownerReferences. If there is at least 1, the controllers do not create the Cluster ownerReference.

The reason I'd like to always have a Cluster ownerReference is so we can identify all of a Cluster's dependent objects and issue delete requests for each dependent object when a Cluster is deleted. This helps address #985 along with #1180

**Environment:**

- Cluster-api version: v0.1.8 and earlier
- Minikube/KIND version:
- Kubernetes version: (use `kubectl version`):
- OS (e.g. from `/etc/os-release`):

### PR DESCRIPTION
**Pull request by liztio**
**What this PR does / why we need it**:
Some pieces of cluster infrastructure (i.e. VPCs for CAPA)  cannot be successfully deleted until after after the machines in the cluster are all deleted. This pull requests blocks cluster deletion until all owned resources are deleted.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Part of #985 but it likely won't fix it until/unless we fix #1190 

**Special notes for your reviewer**:
This _only_ covers Machines, MachineSets, and MachineDeployments

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**Release note**:

```release-note
Deleting a cluster will block until all owned MachineDeployments, MachineSets, and Machines are deleted. 
```

### CODE REVIEW

File: pkg/controller/cluster/cluster_controller.go [hunk 1]
```diff
@@ -168,3 +199,61 @@ func (r *ReconcileCluster) Reconcile(request reconcile.Request) (reconcile.Resul
...
+	var (
+		deployments clusterv1.MachineDeploymentList
+		sets        clusterv1.MachineSetList
+		machines    clusterv1.MachineList
+		children    []runtime.Object
+	)
+
+	ns := cluster.GetNamespace()
+
+	err := r.List(ctx, client.InNamespace(ns), &deployments)
```
[the remark this answers is not in the document]
Comment (vincepri) [reply pkg/controller/cluster/cluster_controller.go]: We can use the cluster label filter to get objects associated with a cluster

File: pkg/controller/cluster/cluster_controller.go [hunk 2]
```diff
@@ -123,6 +124,36 @@ func (r *ReconcileCluster) Reconcile(request reconcile.Request) (reconcile.Resul
...
+					return reconcile.Result{}, errors.Wrapf(err, "couldn't create accessor for %T", child)
+				}
+
+				if accessor.GetDeletionTimestamp() != nil {
+					continue
+				}
+
+				gvk := child.GetObjectKind().GroupVersionKind().String()
+
+				klog.V(4).Infof("deleting cluster %s: Deleting %s %s", name, gvk, accessor.GetName())
```
Comment (vincepri) [inline pkg/controller/cluster/cluster_controller.go]: Log messages should start with uppercase letter

File: pkg/controller/cluster/cluster_controller.go [hunk 3]
```diff
@@ -123,6 +124,36 @@ func (r *ReconcileCluster) Reconcile(request reconcile.Request) (reconcile.Resul
...
 		}
 
+		children, err := r.listChildren(context.Background(), cluster)
+		if err != nil {
+			klog.Infof("Failed to list dependent objects of cluster %s/%s: %v", cluster.ObjectMeta.Namespace, cluster.ObjectMeta.Name, err)
+			return reconcile.Result{}, err
+		}
+
+		if len(children) > 0 {
+			klog.Infof("deleting cluster %s: %d children still exist, will requeue", name, len(children))
```
Comment (vincepri) [inline pkg/controller/cluster/cluster_controller.go]: Same here, uppercase

File: pkg/controller/cluster/cluster_controller.go [hunk 4]
```diff
@@ -123,6 +124,36 @@ func (r *ReconcileCluster) Reconcile(request reconcile.Request) (reconcile.Resul
...
+
+				gvk := child.GetObjectKind().GroupVersionKind().String()
+
+				klog.V(4).Infof("deleting cluster %s: Deleting %s %s", name, gvk, accessor.GetName())
+				if err := r.Delete(context.Background(), child, client.PropagationPolicy(metav1.DeletePropagationForeground)); err != nil {
+					return reconcile.Result{}, errors.Wrapf(err, "deleting cluster %s: failed to delete %s %s", name, gvk, accessor.GetName())
+				}
+			}
+
+			return reconcile.Result{Requeue: true}, nil
```
Comment (vincepri) [inline pkg/controller/cluster/cluster_controller.go]: Do we want to requeue immediately, or can we wait 10-20 seconds?

Comment (liztio) [reply pkg/controller/cluster/cluster_controller.go]: Are we worried about overloading something? I figured it would go to the end of the queue so it wouldn't take up space on a busy cluster, and be more responsive on an empty one. Open to discussion though

Comment (vincepri) [reply pkg/controller/cluster/cluster_controller.go]: I figured Machines for example in AWS can take a while to get deleted, I don't have a strong opinion, but seems an immediate requeue might be putting pressure on the API server for little gain

**Comment by ncdc** [comment]
Currently reviewing.

/hold

pending comments

File: pkg/controller/cluster/cluster_controller.go [hunk 5]
```diff
@@ -168,3 +220,83 @@ func (r *ReconcileCluster) Reconcile(request reconcile.Request) (reconcile.Resul
...
+		return r.clusterClient.MachineDeployments(ns).List(m)
+	}
+	sfunc := func(_ context.Context, m metav1.ListOptions) (runtime.Object, error) {
+		return r.clusterClient.MachineSets(ns).List(m)
+	}
+	mfunc := func(_ context.Context, m metav1.ListOptions) (runtime.Object, error) {
+		return r.clusterClient.Machines(ns).List(m)
+	}
+
+	deployments, err := pager.New(dfunc).List(ctx, opts)
```
Comment (ncdc) [inline pkg/controller/cluster/cluster_controller.go]: Thinking more about using `r.clusterClient`... if we use `r.Client.List` instead, it will take advantage of the shared informer's cache for each type. That's probably better. WDYT?

File: pkg/controller/cluster/cluster_controller.go [hunk 6]
```diff
@@ -168,3 +220,83 @@ func (r *ReconcileCluster) Reconcile(request reconcile.Request) (reconcile.Resul
...
+	if err != nil {
+		return []runtime.Object{}, errors.Wrapf(err, "Failed to list MachineDeployments in %s", ns)
+	}
+	dlist, ok := deployments.(*clusterv1.MachineDeploymentList)
```
[the remark this answers is not in the document]
Comment (ncdc) [reply pkg/controller/cluster/cluster_controller.go]: NOTE: after I wrote this comment, I wrote another one above about using `r.Client` instead. Please see that first. Thanks.

File: pkg/controller/cluster/cluster_controller_test.go [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+				pointer.OwnerReferences = append(pointer.OwnerReferences, metav1.OwnerReference{
+					UID:  types.UID(ref),
+					Kind: "Cluster",
+				})
+			}
+
+			result := pointsTo(pointer, target)
+			if result != test.expected {
+				t.Errorf("expected %v, got %v", test.expected, result)
+
```
Comment (ncdc) [inline pkg/controller/cluster/cluster_controller_test.go]: Please remove this blank line

File: pkg/controller/cluster/cluster_controller_test.go [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+		{
+			name:     "multiple refs one right",
+			refIDs:   []string{"m4g1c", targetID},
+			expected: true,
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+
```
Comment (ncdc) [inline pkg/controller/cluster/cluster_controller_test.go]: Please remove this blank line

File: pkg/controller/cluster/cluster_controller.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (liztio) [reply pkg/controller/cluster/cluster_controller.go]: I've tested this code and it _does_ work as intended on current capa. But I'll update this anyway

File: pkg/controller/cluster/cluster_controller.go (hunk 5 above)
Comment (liztio) [reply pkg/controller/cluster/cluster_controller.go]: [per Slack discussion](https://kubernetes.slack.com/archives/C8TSNPY4T/p1563904482131800?thread_ts=1563904056.129800&cid=C8TSNPY4T), we've decided to keep the clusterClient for easier use of the pager

File: pkg/controller/cluster/cluster_controller.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (ncdc) [reply pkg/controller/cluster/cluster_controller.go]: I'm really surprised this type assertion works...

File: pkg/controller/cluster/cluster_controller.go [hunk 9]
```diff
@@ -151,20 +204,99 @@ func (r *ReconcileCluster) Reconcile(request reconcile.Request) (reconcile.Resul
...
+
+	for _, d := range dlist.Items {
+		if pointsTo(&d.ObjectMeta, &cluster.ObjectMeta) {
+			children = append(children, d.DeepCopyObject())
+		}
+	}
+
+	for _, s := range slist.Items {
+		if pointsTo(&s.ObjectMeta, &cluster.ObjectMeta) {
+			children = append(children, s.DeepCopyObject())
```
Comment (ncdc) [inline pkg/controller/cluster/cluster_controller.go]: I wonder if we should only append MachineSets if they do **not** belong to a MachineDeployment?

File: pkg/controller/cluster/cluster_controller.go [hunk 10]
```diff
@@ -151,20 +204,99 @@ func (r *ReconcileCluster) Reconcile(request reconcile.Request) (reconcile.Resul
...
+		}
+	}
+
+	for _, m := range mlist.Items {
+		if pointsTo(&m.ObjectMeta, &cluster.ObjectMeta) {
+			children = append(children, m.DeepCopyObject())
```
Comment (ncdc) [inline pkg/controller/cluster/cluster_controller.go]: I wonder if we should only append Machines if they do **not** belong to a MachineSet?

File: pkg/controller/cluster/cluster_controller.go [hunk 11]
```diff
@@ -119,18 +142,48 @@ func (r *ReconcileCluster) Reconcile(request reconcile.Request) (reconcile.Resul
...
+		children, err := r.listChildren(context.Background(), cluster)
+		if err != nil {
+			klog.Infof("Failed to list dependent objects of cluster %s/%s: %v", cluster.ObjectMeta.Namespace, cluster.ObjectMeta.Name, err)
+			return reconcile.Result{}, errors.WithStack(err)
+		}
+
+		if len(children) > 0 {
+			klog.Infof("Deleting cluster %s: %d children still exist, will requeue", name, len(children))
+			for _, child := range children {
+
```
Comment (ncdc) [inline pkg/controller/cluster/cluster_controller.go]: Please remove this blank line

File: pkg/controller/cluster/cluster_controller.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (liztio) [reply pkg/controller/cluster/cluster_controller.go]: it's probably because I explicitly call the typed client functions, it doesn't need to do a conversion

[the remark this answers is not in the document]
Comment (ncdc) [reply pkg/controller/cluster/cluster_controller.go]: Ah, I see what's happening. If there is no pagination (continue = 0), it returns the result from the ListPageFunc (where you're using the typed clients). However, if there _is_ pagination, that's when it returns the other list type.

I think you'll need to use `meta.ExtractList` to be confident this works in all cases. If you want to test your code out without it to see if it breaks, you can store `pager.New()` as a variable, then set its `PageSize` to 1. Assuming you have >1 deployments, sets, or machines, this should paginate for you.

[the remark this answers is not in the document]
Comment (liztio) [reply pkg/controller/cluster/cluster_controller.go]: I ended up replacing this with meta.EachListFunc(), and it cleaned up the code a bit!

File: pkg/controller/cluster/cluster_controller.go (hunk 9 above)
Comment (liztio) [reply pkg/controller/cluster/cluster_controller.go]: MachineSets created by a deployment are owned by that deployment, not the cluster: 

```yaml
    ownerReferences:
    - apiVersion: cluster.k8s.io/v1alpha1
      blockOwnerDeletion: true
      controller: true
      kind: MachineDeployment
      name: whinnyapolis-machinedeployment
      uid: 44b5c291-ad90-11e9-b9d6-0242ac110003
    resourceVersion: "33872"
```

Comment (ncdc) [reply pkg/controller/cluster/cluster_controller.go]: So, there is potentially a bug in release-0.1 where the MachineSet controller is trying to add an ownerRef to the cluster, but it only does it if there are 0 ownerRefs. We identified similar logic in another controller either while triaging a bug, or working on alpha2 (I don't remember which), and I believe we decided that the length check for 0 ownerRefs was incorrect.

I think given this logic is in place in both the MachineDeployment and MachineSet controllers, your logic is ok. If, however, we ever change the logic and instead decide we want all MachineDeployments and MachineSets to have an ownerRef pointing back at the cluster, we should probably adjust your logic here so we don't add MachineSets that belong to MachineDeployments, or Machines that belong to MachineSets. Does that make sense?

Comment (ncdc) [reply pkg/controller/cluster/cluster_controller.go]: I've been thinking about this more and I think I need to revise my comment 😄.

tl;dr: we should ensure that all MachineDeployments, MachineSets, and Machines have a Cluster owner reference; otherwise, deleting a Cluster could miss deleting some objects, and we remain in this stuck state.

Longer version:

I think we may want to update the MachineDeployment controller to remove the length=0 check when adding the owner reference to the cluster. Otherwise, if for some reason a MachineDeployment already has an owner reference to something else (i.e. a user manually set it to something), the MachineDeployment will not have the cluster listed as an owner. When someone deletes the cluster, it won't trigger this logic to delete the MachineDeployment, which could give us the same result we have now - the cluster won't actually get deleted and things will be stuck running in circles.

The same comment applies to the MachineSet and Machine controllers.

WDYT @detiber @vincepri @chuckha

**Comment by liztio** [comment]
well it passes now :woman_shrugging:

File: pkg/controller/cluster/cluster_controller.go [hunk 12]
```diff
@@ -119,18 +141,47 @@ func (r *ReconcileCluster) Reconcile(request reconcile.Request) (reconcile.Resul
...
+			klog.Infof("Failed to list dependent objects of cluster %s/%s: %v", cluster.ObjectMeta.Namespace, cluster.ObjectMeta.Name, err)
+			return reconcile.Result{}, errors.WithStack(err)
+		}
+
+		if len(children) > 0 {
+			klog.Infof("Deleting cluster %s: %d children still exist, will requeue", name, len(children))
+			for _, child := range children {
+				accessor, err := meta.Accessor(child)
+				if err != nil {
+					return reconcile.Result{}, errors.Wrapf(err, "couldn't create accessor for %T", child)
```
Comment (ncdc) [inline pkg/controller/cluster/cluster_controller.go]: Should we aggregate all errors in this for loop so we can try to delete as much as we can?

File: pkg/controller/cluster/cluster_controller.go [hunk 13]
```diff
@@ -151,20 +202,91 @@ func (r *ReconcileCluster) Reconcile(request reconcile.Request) (reconcile.Resul
...
+	}
+
+	eachFunc := func(o runtime.Object) error {
+		acc, err := meta.Accessor(o)
+		if err != nil {
+			return err
+		}
+
+		if pointsTo(acc.GetOwnerReferences(), &cluster.ObjectMeta) {
+			children = append(children, o.DeepCopyObject())
```
Comment (ncdc) [inline pkg/controller/cluster/cluster_controller.go]: I don't think we need to deep copy the child?

File: pkg/controller/cluster/cluster_controller.go (hunk 12 above)
Comment (liztio) [reply pkg/controller/cluster/cluster_controller.go]: The deletion happens in a later step, I think Accessor is unlikely to fail. But it might be a good idea to aggregate the errors for the actual deletion loop

File: pkg/controller/cluster/cluster_controller.go [hunk 14]
```diff
@@ -119,18 +142,60 @@ func (r *ReconcileCluster) Reconcile(request reconcile.Request) (reconcile.Resul
...
+
+				klog.V(4).Infof("Deleting cluster %s: Deleting %s %s", name, gvk, accessor.GetName())
+				if err := r.Delete(context.Background(), child, client.PropagationPolicy(metav1.DeletePropagationForeground)); err != nil {
+					err = errors.Wrapf(err, "deleting cluster %s: failed to delete %s %s", name, gvk, accessor.GetName())
+					klog.Infof("deletion error: %v", err)
+					errList = append(errList, err)
+				}
+			}
+
+			if len(errList) == 1 {
```
Comment (ncdc) [inline pkg/controller/cluster/cluster_controller.go]: The Aggregate code handles array lengths of 0, 1, and 2+ correctly - you can safely replace it with a check for > 0.

File: pkg/controller/cluster/cluster_controller.go (hunk 12 above)
Comment (ncdc) [reply pkg/controller/cluster/cluster_controller.go]: I agree that Accessor is unlikely to fail, but I'd still like to include it in errList instead of returning early.

File: pkg/controller/cluster/cluster_controller.go [hunk 15]
```diff
@@ -119,18 +142,60 @@ func (r *ReconcileCluster) Reconcile(request reconcile.Request) (reconcile.Resul
...
 		if !util.Contains(cluster.ObjectMeta.Finalizers, clusterv1.ClusterFinalizer) {
-			klog.Infof("reconciling cluster object %v causes a no-op as there is no finalizer.", name)
+			klog.Infof("Reconciling cluster object %v causes a no-op as there is no finalizer.", name)
 			return reconcile.Result{}, nil
 		}
 
-		klog.Infof("reconciling cluster object %v triggers delete.", name)
+		children, err := r.listChildren(context.Background(), cluster)
+		if err != nil {
+			klog.Infof("Failed to list dependent objects of cluster %s/%s: %v", cluster.ObjectMeta.Namespace, cluster.ObjectMeta.Name, err)
```
Comment (ncdc) [inline pkg/controller/cluster/cluster_controller.go]: Let's do Errorf instead of Infof

File: pkg/controller/cluster/cluster_controller.go [hunk 16]
```diff
@@ -119,18 +142,60 @@ func (r *ReconcileCluster) Reconcile(request reconcile.Request) (reconcile.Resul
...
+			return reconcile.Result{}, errors.WithStack(err)
```
Comment (ncdc) [inline pkg/controller/cluster/cluster_controller.go]: I don't think we need WithStack and instead we can return err directly because r.listChildren is our function, and it properly wraps/stacks all the errors it returns. WDYT?

File: pkg/controller/cluster/cluster_controller.go [hunk 17]
```diff
@@ -119,18 +142,60 @@ func (r *ReconcileCluster) Reconcile(request reconcile.Request) (reconcile.Resul
...
+				if accessor.GetDeletionTimestamp() != nil {
+					continue
+				}
+
+				gvk := child.GetObjectKind().GroupVersionKind().String()
+
+				klog.V(4).Infof("Deleting cluster %s: Deleting %s %s", name, gvk, accessor.GetName())
+				if err := r.Delete(context.Background(), child, client.PropagationPolicy(metav1.DeletePropagationForeground)); err != nil {
+					err = errors.Wrapf(err, "deleting cluster %s: failed to delete %s %s", name, gvk, accessor.GetName())
+					klog.Infof("deletion error: %v", err)
```
Comment (ncdc) [inline pkg/controller/cluster/cluster_controller.go]: Let's make this Errorf instead of Infof, and please capitalize Deletion

File: pkg/controller/cluster/cluster_controller.go (hunk 16 above)
Comment (liztio) [reply pkg/controller/cluster/cluster_controller.go]: I'll be honest, I have no idea what WithStack does, I added it because you suggested it a few blocks up

Comment (liztio) [reply pkg/controller/cluster/cluster_controller.go]: I guess you only need to add it once?

Comment (ncdc) [reply pkg/controller/cluster/cluster_controller.go]: By default, errors in Go do not include any stack trace information. Without it, it makes it very difficult to determine the source of an error (i.e. source code file & line), especially if the same error can be created in multiple places. github.com/pkg/errors makes it much easier to embed that stack information, and it includes a special custom format printer to print out the stack trace information.

The general guidance I've used in the past is that every line of code you write in a project that returns an error must either already have stack info, or you must wrap it. For first class functions (those defined in your project), you can/should assume that any errors they return are already wrapped. When you are invoking functions from vendored libraries, you can either check to see if they use github.com/pkg/errors, or you can just wrap.

File: pkg/controller/cluster/cluster_controller.go [hunk 18]
```diff
@@ -119,18 +142,57 @@ func (r *ReconcileCluster) Reconcile(request reconcile.Request) (reconcile.Resul
...
+				if accessor.GetDeletionTimestamp() != nil {
+					continue
+				}
+
+				gvk := child.GetObjectKind().GroupVersionKind().String()
+
+				klog.V(4).Infof("Deleting cluster %s: Deleting %s %s", name, gvk, accessor.GetName())
+				if err := r.Delete(context.Background(), child, client.PropagationPolicy(metav1.DeletePropagationForeground)); err != nil {
+					err = errors.Wrapf(err, "deleting cluster %s: failed to delete %s %s", name, gvk, accessor.GetName())
+					klog.Errorf("deletion error: %v", err)
```
Comment (ncdc) [inline pkg/controller/cluster/cluster_controller.go]:
```suggestion
					klog.Errorf("Deletion error: %v", err)
```

File: pkg/controller/cluster/cluster_controller.go (hunk 12 above)
Comment (ncdc) [reply pkg/controller/cluster/cluster_controller.go]: I think this is my only remaining comment

**Comment by ncdc** [comment]
1 last comment, then LGTM. I'd like 1 more review:
/assign @vincepri

File: pkg/controller/cluster/cluster_controller.go [hunk 19]
```diff
@@ -151,20 +213,91 @@ func (r *ReconcileCluster) Reconcile(request reconcile.Request) (reconcile.Resul
...
+	if err := meta.EachListItem(sets, eachFunc); err != nil {
+		return []runtime.Object{}, errors.Wrapf(err, "couldn't iterate MachineSets for cluster %s/%s", ns, cluster.Name)
+	}
+
+	machines, err := pager.New(mfunc).List(ctx, opts)
+	if err != nil {
+		return []runtime.Object{}, errors.Wrapf(err, "failed to list Machines in %s/%s", ns, cluster.Name)
+	}
+	if err := meta.EachListItem(machines, eachFunc); err != nil {
+		return []runtime.Object{}, errors.Wrapf(err, "couldn't iterate Machines for cluster %s/%s", ns, cluster.Name)
```
Comment (vincepri) [inline pkg/controller/cluster/cluster_controller.go]: Can we return nil instead of an empty slice in when we have errors?

File: pkg/controller/cluster/cluster_controller.go [hunk 20]
```diff
@@ -151,20 +213,91 @@ func (r *ReconcileCluster) Reconcile(request reconcile.Request) (reconcile.Resul
...
+	}
+
+	return children, nil
+}
+
+func pointsTo(refs []metav1.OwnerReference, target *metav1.ObjectMeta) bool {
```
Comment (vincepri) [inline pkg/controller/cluster/cluster_controller.go]: We could move this to the util package, others might find this useful as well

**Comment by vincepri** [comment]
Mostly nits, which can be fixed later. 

/lgtm
/hold

Feel free to unhold when ready

File: pkg/controller/cluster/cluster_controller.go (hunk 19 above)
Comment (liztio) [reply pkg/controller/cluster/cluster_controller.go]: huh, TIL

File: pkg/controller/cluster/cluster_controller.go [hunk 21]
```diff
@@ -151,20 +213,81 @@ func (r *ReconcileCluster) Reconcile(request reconcile.Request) (reconcile.Resul
...
+
+		return nil
+	}
+
+	deployments, err := pager.New(dfunc).List(ctx, opts)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to list MachineDeployments in %s/%s", ns, cluster.Name)
+	}
+	if err := meta.EachListItem(deployments, eachFunc); err != nil {
+		return []runtime.Object{}, errors.Wrapf(err, "couldn't iterate MachinesDeployments for cluster %s/%s", ns, cluster.Name)
```
Comment (ncdc) [inline pkg/controller/cluster/cluster_controller.go]:
```suggestion
		return nil, errors.Wrapf(err, "couldn't iterate MachinesDeployments for cluster %s/%s", ns, cluster.Name)
```

**Comment by ncdc** [comment]
/lgtm
/approve

@liztio please remove the hold after you retest against CAPA.

**Comment by liztio** [comment]
```
I0724 18:48:29.221278       1 scope.go:219] [cluster-actuator]/cluster.k8s.io/v1alpha1/default/whinnyapolis "level"=1 "msg"="updating cluster status"  
I0724 18:48:29.314440       1 cluster_controller.go:196] Cluster object whinnyapolis deletion successful, removing finalizer.
```
